### ISSUE TITLE
[Feature Request] Windows build

### ISSUE THREAD

**Issue by spidey** [adrienverge/openfortivpn#169]
Would it be possible to port this to Windows?

**Comment by mrbaseman** [comment]
I have tried to build _openfortivpn_ with [mingw](http://www.mingw.org/) and with [cygwin](https://www.cygwin.com/), but didn't succeed with either option without modifying the source. The code makes use of a few parts of the linux kernel, like network interfaces, routing tables, parallel threads,...
The kernel used in Mac OSX also is a unix-like one, but for the Mac OSX port of _openfortivpn_ a few changes to the source code were needed, involving workarounds like parsing the output of `netstat` instead of reading entries of the `/proc` filesystem. It might be possible to find similar workarounds to port the code to cygwin for instance. 
The next obstacle is that _openfortivpn_ calls pppd for handling the vpn connection, but there seems to be [no pppd for cygwin yet](https://cygwin.com/ml/cygwin/2003-01/msg01166.html), so part of the port would be porting pppd, or at least finding a replacement for the features of pppd we use. I guess that writing a standalone ssl vpn application based on [windows sockets](https://technet.microsoft.com/en-us/library/ms740139.aspx) and maybe AnyTun as suggested [here](https://stackoverflow.com/questions/8525577/virtual-network-interface-for-windows) would probably not be much more effort.
Anyhow, if anyone successfully ports _openfortivpn_ to windows I would be happy to review a pull request.

**Comment by DimitriPapadopoulos** [comment]
A substantial part of the code is based on POSIX system calls. Porting _openfortivpn_ means porting or even rewriting at least these files: `config.c`, `io.c`, `ipv4.c`, `log.c`, `tunnel.c`, `userinput.c`.

A supported proprietary client is already available on Windows. The Windows client uses primarily IPSec, but can fall back to SSL. What value do you see value in porting _openfortivpn_ to Windows? I'm not against it, just pointing out it is a significant project...

**Comment by mrbaseman** [comment]
An argument for the windows port would be to have a more or less unique client that supports all major operating systems on the market, which makes support for all user groups a bit easier.

**Comment by spidey** [comment]
Yeah, after opening the issue I thought for a while and realized how dependent on the OS this would be.

I was hoping to get a decent client for Windows as well, as the proprietary one keeps opening warning dialogs and disconnecting every half an hour or so.

**Comment by spidey** [comment]
Should I close the issue then?

**Comment by DimitriPapadopoulos** [comment]
Well, it's not impossible to port _openfortivpn_ - just complicated!

By the way the FortiClient software used to work well for me, it used to operate for hours without disconnection. Perhaps something's wrong on the Fortinet device or the Windows computer? Note that there are [alternative clients for Windows](https://serverfault.com/questions/539715/alternative-to-forticlient-for-ipsec-vpn#558817). Also if your IT team supports Windows only, they might have disabled VPN SSL on the Fortinet device - this had been reported in another thread of discussion.

**Comment by spidey** [comment]
Since porting to Windows wouldn't be a simple task, involving porting dependencies as well, I'm closing the issue. Thank you on the feedback and suggestions on alternatives.

**Comment by evandrix** [comment]
is it possible to export FortiVPN into an OpenVPN .ovpn config file sans the 2FA/OTP part, like many other VPN vendors?

**Comment by DimitriPapadopoulos** [comment]
@evandrix This issue has been closed for more than two years and I don't see how your question is related to it. Also is your question related to _openfortivpn_?

**Comment by abubelinha** [comment]
> A supported proprietary client is already available on Windows. The Windows client uses primarily IPSec, but can fall back to SSL. What value do you see value in porting _openfortivpn_ to Windows? I'm not against it, just pointing out it implies significant work...

Value:

1. The propietary client (FortiClient VPN) is just supported in its commercial version
2. The free client version does only maintain the VPN for a limited time. Then you have to restart it again.
3. But the really worst thing is it randomly fails to restart again the VPN. Even if you follow [their recommendations](https://pub.kb.fortinet.com/Platform/Publishing/694/FD36630_f.1.html) installing a [3rd party patch ](https://my.zebrahost.com/knowledgebase/38/Error-code--20199-FortiClient-VPN.html)to solve it, after a few days the fail comes back again randomly. A nightmare.

So, in summary, there is not a free and working windows client.

**Comment by DimitriPapadopoulos** [comment]
I'm not against porting _openfortivpn_ to Windows. It's just too complicated in the short term and there are lots of more urgent tasks that need to be addressed. I just don't see it happening unless someone else commits substantial work.

It will get easier to port to Windows with time, as other changes are hopefully applied:
* directly use PPP code in _openfortivpn_ instead of forking `pppd`,
* run scripts to set routing and name resolution instead of using C code.

**Comment by abubelinha** [comment]
> I'm not against porting openfortivpn to Windows

Yes I know. 

I was just giving you strong reasons to keep the issue open

**Comment by DimitriPapadopoulos** [comment]
Keeping issues that we cannot address in the next few years open won't help in any way. If anyone is interested contributing a Windows version, it's probably best to open issues to discuss specific portability issues and address each one of them.

**Comment by abubelinha** [comment]
OK, sorry, I am not a developer ... maybe I just don't understand how github issues work (I have not clear what _Open / Closed_ means in this context).

**Comment by bertogravscale** [comment]
did you tried to use wsl?

**Comment by abubelinha** [comment]
No. Actually I gave up because they changed from Fortinet to a different VPN. So I don't need openfortivpn any more
